bart/sched/pelt.py

```diff
@@ -0,0 +1,492 @@
+
+        def _to_pelt_samples(time_us):
+            return float(time_us)/Simulator._sample_us
+
+        # Compute max value
+        max_pelt  = (1. - pow(self._geom_y, _to_pelt_samples(task.run_us)))
+        max_pelt /= (1. - pow(self._geom_y, _to_pelt_samples(task.period_us)))
+
+        # Compute min value, by decaying the maximum value
+        min_pelt  = max_pelt
+        min_pelt *= pow(self._geom_y, _to_pelt_samples(task.idle_us))
+
+        return PELTRange(Simulator._signal_max * min_pelt,
+                         Simulator._signal_max * max_pelt)
+
+
+    def getSignal(self, task, start_s=0, end_s=10):
+        """Compute the PELT signal for the specified task and interval.
+
+        :param task: the task we want to simulate the PELT signal for, by
+                     default is the task used to intialize the Simulator
+                     (when specified)
+        :type  task: PeriodicTask
+
+        :param start_s: the start time in [s]
+        :type  start_s: float
+
+        :param end_s: the end time in [s]
+        :type  end_s: float
+
+        :return: :mod:`pandas.DataFrame` instance which reports the computed
+        PELT values at each PELT sample interval. The returned columns are:
+        - Time          : the PELT sample time
+        - PELT_Interval : the PELT sample number
+        - Running       : a boolean reporting if the task was RUNNING in that
+                          PELT interval
+        - pelt_value      : the compute PELT signal value at the end of that PELT
+                          interval
+        """
+
+        # Computed PELT samples
+        samples = []
+        # Decay capping support
+        running_prev = True
+        capping = None
+
+        # Validate input parameters
+        if not isinstance(task, PeriodicTask):
+            raise ValueError("Wrong type for task parameter")
+
+        # Intervals (in PELT samples) for the signal to compute
+        self.start_us = _s_to_us(start_s, Simulator._sample_us)
+        self.end_us = _s_to_us(end_s, Simulator._sample_us)
+
+        # Add initial value to the generated signal
+        pelt_value = self.init_value
+        t_us = self.start_us
+        sample = (_us_to_s(t_us), 0, False, pelt_value)
+        samples.append(sample)
+
+        # Compute following PELT samples
+        t_us = self.start_us + Simulator._sample_us
+        while t_us <= self.end_us:
+
+            # Check if the task was running in the current PELT sample
+            running = task.isRunning(_us_to_ms(t_us))
+
+            # Keep track of sleep start and decay capping time
+            if self.decay_cap_ms and running_prev and not running:
+                capping = t_us + _ms_to_us(self.decay_cap_ms,
+                                           Simulator._sample_us)
+
+            # Assume the task was running for all the current PELT sample
+            active_us = Simulator._sample_us if running else 0
```

Comment:
Just realised: Should use `self._sample_us` here and elsewhere, then it can be overridden by subclasses or specific instances (also then no explicit references to the class name in the body). No need to set it in `__init__`, attribute lookups fall back to class attributes if there's no instance attribute.